### ISSUE TITLE
SPIKE: Select 2-4 pretrained, off the shelf models from MTEB

### ISSUE THREAD

**Issue by bamader** [CDCgov/dibbs-text-to-code#57]
- To get a decent result with our fine-tuning, we need to start with a language model that’s already been trained on a large amount of English data. Our domain is about learning medical semantics and clinical contexts, so having a model that already understands sentential syntax, structure, and high-level lexical nuance is important.
- The Massive Text Embedding Leaderboard (MTEB) on HuggingFace lists the current best open source models at performing a variety of NLP tasks. The charts also show the scores of the models, their size in terms of number of parameters, and their compatibility with libraries like sentence transformers.
- The scope of this ticket is to select two types of model from MTEB:
  - An embedding model meant for semantic search (that task might also go by titles like fill-mask, similarity scoring, and feature extraction)
  - A cross-encoder or reranker model meant for evaluating precise scores between small sets of candidates
- For each type of model, document some considerations (model size, task the model is optimized for, level of domain specificity, etc.) and select 2-4 candidates worth exploring in the next phase. Try to get a mixture of models on various dimensions, most importantly some models that are truly generic (E5, Q3, BGE) and some models that might be clinically optimized (PubMedBert).
- Reference [5] could provide inspiration for additional worthwhile embedding models to use, given the point of that study was looking for high-accuracy generalist models

[5] Excoffier, Jean-Baptiste, et al. “Generalist Embedding Models Are Better At Short-Context Clinical Semantic Search Than Specialized Embedding Models.” Kaduceo, Toulouse, France. 2024. https://arxiv.org/pdf/2401.01943

**Comment by bamader** [comment]
And here's some options for the clinical / specialist domain medical embedding models. Worth getting a benchmark from them even if we hypothesize that generalist that goes domain adapted will do better. NOTE: there are unfortunately some limitations with existing clinical embedders: many of them aren't compiled with a library called `SafeTensors`. This isn't bad on its own, but it creates a serious security vulnerability in `pytorch`'s underlying `tensor.load()` function that prevents the model from loading on a lot of older Intel chips, including Macs before 2021. For security reasons and compatibility I'm going to try to find only models that are SafeTensor compatible:

PubMedBert Base Embeddings: https://huggingface.co/NeuML/pubmedbert-base-embeddings; pretty straightforward model, picks up clinical context from medical abstracts. Suited to the domain, but not specifically task trained.

Gemma300: https://huggingface.co/sentence-transformers/embeddinggemma-300m-medical; A model I am very excited to investigate. Specifically developed for/with sentence transformers using a new google architecture for medical information retrieval, I think this is probably a best in slot example of "cutting edge" for pretrained clinical embeddings.

CORe Clinical Outcome BioBert: https://huggingface.co/bvanaken/CORe-clinical-outcome-biobert-v1; an interesting model proposition that starts with bio bert and then further pre-trains on clinical notes for additional adaptation. This model is a case where things might be _too_ pretrained for us to do much more, but it's an important benchmark case study.

Clinical Longformer: https://huggingface.co/Gaborandi/Clinical-Longformer-MLM-pubmed; the clinical longformer model is a competitor to BioBert, and this one is further pre-trained on eCR formatted data extraction from the MIMIC-III dataset (a very good one for eCR refinement) and PubMed abstracts

Bioformer: bioformers/bioformer-8L; Bioformer is an interesting model trained _only_ on domain relevant data. Unlike other models which learn context from general or full text English, Bioformer learns everything in a clinical lens. The impact of that is worth playing with. Also available as a 16-layer model with smaller embedding size; the 16-layer should give better performance at the cost of some speed.

Cambridge's BioRedditBert: https://huggingface.co/cambridgeltl/BioRedditBERT-uncased; yeah I know how it sounds, but the dataset coverage and additional context training on human-produced text is what's interesting to me. The way people talk on social media has interesting implications for how a model might pick up context, particularly around different semantics or substitutions of referring to the same thing.

**Comment by bamader** [comment]
Models that were considered but which have the security vulnerability described above:

* BioClinicalBert
* Clinical Bert
* Microsoft BiomedNLP BiomedBert
* BlueBert and variants (e.g. BlueBert PubMed MIMIC)
* Bioclinical Modern Bert
